app/contacts/contact_ftp.py

```diff
@@ -0,0 +1,217 @@
+
+            await self.handle_agent_file(name, bytes_obj, r_class)
+            connection.response("226", "data transfer done")
+            del stream
+            return True
+
+        real_path, virtual_path = self.get_paths(connection, rest)
+        name = str(virtual_path).split("/")
+        self.logger.debug("File received: %s" % str(name[-1]))
+        r_class = CalderaServer(self.contact_svc, self.file_svc, self.logger, self.directory)
+
+        if await connection.path_io.is_dir(real_path.parent):
+            coro = stor_worker(self, connection, rest)
+            task = asyncio.create_task(coro)
+            connection.extra_workers.add(task)
+            code, info = "150", "data transfer started"
+        else:
+            code, info = "550", "path unreachable"
+        connection.response(code, info)
+        return True
+
+    async def handle_agent_file(self, file_name, file_bytes, r_class):
+        if re.match(r"^Alive\.txt$", file_name[-1]):
+            profile = json.loads(file_bytes.decode())
+            paw, response = await r_class.create_beacon_response(profile)
+            success = r_class.write_beacon_response_file(paw, response)
+            if not success:
+                self.logger.debug("ERROR: Failed to create response")
+
+        elif re.match(r"^Payload\.txt$", file_name[-1]):
+            profile = json.loads(file_bytes.decode())
+            file_path, contents, display_name = await r_class.get_payload_file(profile)
+            if file_path is not None:
+                r_class.write_payload_file(profile.get('paw'), profile.get('file'), str(contents))
+
+        else:
+            paw = file_name[-2]
+            filename = file_name[-1]
+            await r_class.submit_uploaded_file(paw, filename, file_bytes)
+
+
+class CalderaServer:
+    def __init__(self, contact, file, log, directory):
+        self.contact_svc = contact
+        self.file_svc = file
+        self.logger = log
+        self.home = os.getcwd()
+        self.directory = directory
+
+    async def create_beacon_response(self, profile):
+        paw = profile.get('paw')
+        profile['paw'] = paw
+        profile['contact'] = profile.get('contact', 'ftp')
+        agent, instructions = await self.contact_svc.handle_heartbeat(**profile)
+        response = dict(paw=agent.paw,
+                        sleep=await agent.calculate_sleep(),
+                        watchdog=agent.watchdog,
+                        instructions=json.dumps([json.dumps(i.display) for i in instructions]))
+        if agent.pending_contact != agent.contact:
+            response['new_contact'] = agent.pending_contact
+            self.logger.debug('Sending agent instructions to switch from C2 channel %s to %s'
+                              % (agent.contact, agent.pending_contact))
+        return agent.paw, response
+
+    def write_beacon_response_file(self, paw, response):
+        filename = str(self.home + self.directory)
```

Comment:
if you're worried about passing slashes back and for, you can use `os.path.join`